lib/searches.js

```diff
@@ -1,252 +1,343 @@
-        console.log("Link <title>", sourceTitle);
+        let title = $("title").text();
+        console.log("Link <title>", title);
+        
+        // Insert link into temp edges table
+        dbAPI.insertIntoTempEdges(searchID, randomLink.SourceURL, randomLink.SourceTitle, randomLink.DestinationURL, title).catch((err) => {
+            console.log(err);
+        });
+        
+        // Written by Nigel. Find keyword if present
+        if ((keyword !== "") && (body.indexOf(keyword) > -1)) {
+            // We found the search term: we're done.
+            result.status = TERM_FOUND;
+            result.keywordURL = randomLink.DestinationURL;
+            return;
+        } 
+        
         let links = $("a");
-
-        // Loop through random page links until an absolute link that looks valid is found.
-        while (!isValid){
-            if (((typeof(randomUrl) == "undefined" || !validUrl.isUri(randomUrl)) || !r.test(randomUrl)) && numInvalidLinks < 20) {
-                randomUrl = $($(links).get(Math.floor(Math.random() * $(links).length))).attr("href");
-                numInvalidLinks++;
-                continue;
+        let numLinks = 0;
+        
+        // Loop through page links and find valid-looking links
+        $(links).each(function(i, link){
+            let linkURL = $(link).attr("href");
+            
+            // Try to resolve relative links with host name
+            if (linkURL && !r.test(linkURL)){
+                linkURL = resUrl.resolve(randomLink.DestinationURL, linkURL);
             }
-            // If page loops for valid links 20 times, conclude there are none and return.
-            if(numInvalidLinks == 20){
-                console.log("No more links. Sending results...");
-                return;
+            
+            // If link looks valid, add it to array (up to 50 links)
+            if (linkURL && 
+                validUrl.isUri(linkURL) &&
+                linkURL != "javascript:void(0)" &&
+                linkURL != "javascript:void(0);") {
+                    
+                numLinks++;
+                destLinks.push({
+                    "SourceURL": randomLink.DestinationURL,
+                    "SourceTitle": title,
+                    "DestinationURL": linkURL
+                });
             }
-            // If link looks valid, add to edge list and return.
-            isValid = true;
-            numInvalidLinks = 0;
-            edges[level] = ({
-                "SourceUrl": url,
-                "SourceTitle": sourceTitle,
-                "DestinationUrl": randomUrl,
-                "TermFound": false
-            }); 
+            // Once 50 links that look valid are added, return list
+            if (numLinks >= 50) {
+                return false;
+            } 
+        });
+        
+        return destLinks;
+        
+    }).catch((err) => {
+        // If error loading root, or more than 30 links tried,exit
```

Comment:
Let's handle errors in the same way we do in `getBFSLinks` so we have consistent behavior. Wrap the `then` handler in a try/catch and throw an error which has been marked so that we know it's a code error which is our fault instead of `request`s.
```js
if (err.notFetchError){
    delete err.notFetchError;
    throw error;
} else if (level === 0) {
    // if the root URL is busted, we need to actually reject.
    throw err;
} else if (attempts > 30) {
   return [];
} else {
    return getDFSLinks(sourceLinks, level, searchDepth, searchID, attempts+1, result, keyword);
}
```